### PR TITLE
shielded-pool: delegate supply tracking to components

### LINKED ISSUE #4018: shielded-pool: remove supply tracking from component
We want to remove supply tracking functionality from the shielded pool. In practice, this means delegating accounting (if needed) to components that need to call `mint_note` and removing the `SupplyWrite` trait.

The high-level plan:
- [ ] shielded-pool: remove supply increment from `mint_note`
- [ ] shielded-pool: remove `SupplyWrite` 
- [ ] staking: implement `ValidatorPoolTracker` for delegation tokens
- [ ] ibc: remove redundant supply tracking

We won't need to replace the Ics20 usage of `SupplyWrite` because it was 1/ broken 2/ redundant with the existing value balance rate-limit mechanism.

```rust=

pub trait ValidatorPoolRead: StateRead {
    pub fn get_validator_pool(&self, id: &IdentityKey) -> Option<Value>;
}

pub(crate) ValidatorPoolWrite: StateWrite {
    pub fn increase_validator_pool(&mut self, id: &IdentityKey, value: Value) -> Result<Value>;
    pub fn decrease_validator_pool(&mut self, id: &IdentityKey, value: Value) -> Result<Value>;
}
```

Adding the following state keys:

- `staking/validators/data/pool/balance/{id}` -> Amount
- `staking/chain/total_bonded` -> Amount

### PR DESCRIPTION
**Pull request by erwanor**
Close #4018

### CODE REVIEW

File: crates/core/component/stake/src/state_key.rs [hunk 1]
```diff
@@ -63,9 +63,18 @@ pub mod validators {
             format!("staking/validators/data/power/{id}")
         }
     }
-    pub mod bonding_state {
-        pub fn by_id(id: &crate::IdentityKey) -> String {
-            format!("staking/validators/data/bonding_state/{id}")
```
[the remark this answers is not in the document]
Comment (erwanor) [reply crates/core/component/stake/src/state_key.rs:68]: Spot on, we also have other state breaking changes in here like removing the supply state keys and not writing supply increments to the JMT.

**Comment by erwanor** [comment]
This PR leaves the asset metadata tracking in the shielded pool. The reason is that it is useful to have a single RPC capable of serving queries about asset metadata, and the shielded pool crate is a lower level component that all customers of `register_denom` already depend on (independently), in other words, there isn't a great place to move it to.

**Review by cratelyn — APPROVED** [review]

File: crates/core/component/stake/src/component/validator_handler/validator_store.rs [hunk 2]
```diff
@@ -267,3 +268,64 @@ pub trait ValidatorDataWrite: StateWrite {
...
+
+        if let Some(new_supply) = old_supply.checked_add(&add) {
+            self.put(state_path, new_supply);
+            Some(new_supply)
+        } else {
+            None
+        }
+    }
+
+    /// Checked increase of the validator pool size by the given amount.
```
[the remark this answers is not in the document]
Comment (erwanor) [reply crates/core/component/stake/src/component/validator_handler/validator_store.rs]: Thanks!

File: crates/core/component/shielded-pool/src/component/assets.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+use tracing::instrument;
+
+use crate::state_key;
+
+#[async_trait]
+pub trait AssetRegistryRead: StateRead {
+    async fn denom_by_asset(&self, asset_id: &asset::Id) -> Option<Metadata> {
+        self.get(&state_key::denom_by_asset(asset_id))
+            .await
+            .expect("no deserialization error")
```
Comment (cronokirby) [inline crates/core/component/shielded-pool/src/component/assets.rs:15]: Q: Will this potentially fail for benign resigns we nevertheless want to throw upstream, e.g. wrong storage path? IDK what the convention here is though.

Comment (erwanor) [reply crates/core/component/shielded-pool/src/component/assets.rs:15]: I don't think so, the failure mode for this is if we have a cache miss that hits storage and fails. The most likely reason this happens is some disk failure that cause storage degradation. Either way, it's irrecoverable for callers and the node should halt.

**Comment by erwanor** [comment]
Merging to get out of the way of #4001, happy to revisit.

**Pull request merged**